chunky/src/java/se/llbit/chunky/renderer/InternalRenderManager.java

```diff
@@ -0,0 +1,502 @@
+    };
+  }
+
+  /**
+   * This controls most of the render manager logic.
+   */
+  @Override
+  public void run() {
+    try {
+      while (!isInterrupted()) {
+        // Wait for something to happen
+        ResetReason reason = sceneProvider.awaitSceneStateChange();
+
+        // Copy the new scene state to the bufferedScene
+        final boolean[] sceneReset = {false};
+        synchronized (bufferedScene) {
+          sceneProvider.withSceneProtected(scene -> {
+            if (reason.overwriteState()) {
+              bufferedScene.copyState(scene);
+              sceneReset[0] = true;
+            }
+            if (reason == ResetReason.MATERIALS_CHANGED || reason == ResetReason.SCENE_LOADED) {
+              scene.importMaterials();
+              sceneReset[0] = true;
+            }
+
+            bufferedScene.copyTransients(scene);
+            finalizeAllFrames = scene.shouldFinalizeBuffer();
+            updateRenderState(scene);
+
+            if (reason == ResetReason.SCENE_LOADED) {
+              // Make sure frame is finalized
+              bufferedScene.postProcessFrame(renderTask);
+
+              // Swap buffers so the render canvas will see the current frame.
+              bufferedScene.swapBuffers();
+
+              // Notify the scene listeners.
+              sendSceneStatus(bufferedScene.sceneStatus());
+              canvas.repaint();
+            }
+          });
+        }
+
+        // Select the correct renderer
+        Renderer render = mode == RenderMode.PREVIEW ? getPreviewRenderer() : getRenderer();
+
+        if (sceneReset[0]) {
+          render.sceneReset(this, reason);
+        }
+
+        renderStart = System.currentTimeMillis();
+        if (mode == RenderMode.PREVIEW) {
+          // Preview with no CPU limit
+          pool.setCpuLoad(100);
+          render.setPostRender(previewCallback);
+          render.render(this);
+          pool.setCpuLoad(cpuLoad);
+        } else {
+          // Bail early if render is already done
+          if (bufferedScene.spp > bufferedScene.getTargetSpp()) {
+            sceneProvider.withSceneProtected(scene -> {
+              scene.pauseRender();
+              updateRenderState(scene);
+            });
+          } else if (mode != RenderMode.PAUSED) {
+            render.setPostRender(renderCallback);
+            render.render(this);
+          }
+        }
+
+        if (headless) break;
+      }
+    } catch (InterruptedException e) {
+      // Interrupted
+    } catch (Throwable e) {
+      Log.error("Unchecked exception in render manager.", e);
+    }
+  }
+
+  private Renderer getRenderer() {
+    return renderers.getOrDefault(renderer, EMPTY_RENDERER);
+  }
+
+  private Renderer getPreviewRenderer() {
+    return previewRenderers.getOrDefault(previewRenderer, EMPTY_RENDERER);
+  }
+
+  @Override
+  public String[] getRenderers() {
+    return renderers.keySet().toArray(new String[0]);
+  }
```

Comment:
This could return a `Collection<String>` instead of creating a new array every time.